File: services/realtime-dataSvc.go
package services

import (
    "context"
    "firebase.google.com/go/v4"
    "google.golang.org/api/option"
    "log"
    "ur-admin-backend/utils"
    "fmt"
    "ur-admin-backend/models"
    "time"
    "strconv"
)

var cacheRealtime *models.Cache = models.NewCache()

func GetFirebaseData(tokenName string) (string, error) {
    log.Println("=== GetFirebaseData Start ",tokenName)
   path, error := utils.LoadEnv("credentials")
   if error != nil {
      return "", error
   }

   opt := option.WithCredentialsFile(path)

   url, error := utils.LoadEnv("urldbfirebase")
   if error != nil {
        return "", error   
   }

   if value, ok := cacheRealtime.Get(tokenName); ok {
    log.Println("=== GetFirebaseData Cache ",tokenName)
     return value.(string), nil   
   }

   app, err := firebase.NewApp(context.Background(), &firebase.Config{
     DatabaseURL: url,
   }, opt)
   if err != nil {
      log.Fatalf("init Firebase App: %v\n", err)
   }

    client, err := app.Database(context.Background())
        if err != nil {
                log.Fatalf("getting database client: %v\n", err)
        }   
    ref := client.NewRef("/"+tokenName)
    var data string
    if err := ref.Get(context.Background(), &data); err != nil {
        return "", err
    }

    if len(data) > 0  {
        cacheValue, err := utils.LoadEnv("CACHE_TIME_DATA")
        if(err == nil){
            cacheTime, error := strconv.Atoi(cacheValue)
            if error == nil {
                cacheRealtime.Set(tokenName, data, time.Hour*time.Duration(cacheTime))
            }
        }
        return data, nil
    }
    
    return "", fmt.Errorf("Not found token %s", tokenName) 
}

func ValidateUser(email string, uuidUser string) (bool, error){
   path, error := utils.LoadEnv("credentials")
   if error != nil {
      return false, error
   }
   opt := option.WithCredentialsFile(path)

   url, error := utils.LoadEnv("urldbfirebase")
   if error != nil {
        return false, error   
   }

   key := fmt.Sprintf("%s-%s",email,uuidUser)

    if value, ok := cacheRealtime.Get(key); ok {
    log.Println("=== ValidateUser Cache ",key)
     return value.(bool) , nil   
   }

   app, err := firebase.NewApp(context.Background(), &firebase.Config{
     DatabaseURL: url,
   }, opt)
   if err != nil {
      log.Fatalf("Error inicializando Firebase App: %v\n", err)
   }

    client, err := app.Database(context.Background())
        if err != nil {
                log.Fatalf("error getting database client: %v\n", err)
        }   
    ref := client.NewRef("/users")
    var data []interface{}
    if err := ref.Get(context.Background(), &data); err != nil {
        log.Printf("error getting data: %v\n", err)
        return false, err
    }
    for _, element := range data {
        email := element.(map[string]interface {})["email"]
        uidUser := element.(map[string]interface {})["uidUser"]
        if email == email && uidUser == uuidUser {
            cacheValue, err := utils.LoadEnv("CACHE_TIME_USER")
            if(err == nil){
                cacheTime, error := strconv.Atoi(cacheValue)
                if error == nil {
                    cacheRealtime.Set(key, true, time.Hour*time.Duration(cacheTime))
                }
            }
            return true, nil
        }
    }
    cacheValue,err := utils.LoadEnv("CACHE_TIME_USER_FALSE")
    if(err == nil){
        cacheTime, error := strconv.Atoi(cacheValue)
        if error == nil {
            cacheRealtime.Set(key, false, time.Hour*time.Duration(cacheTime))
        }
    }
    return false , nil
}

File: services/realtimeDataSvc.js
const { Cache } = require('../models/cache');
const { loadEnv } = require('../utils/loadEnv');
const admin = require('firebase-admin');
const fs = require('fs');
const path = require('path');

const cacheRealtime = new Cache();

let app;

function getApp() {
  if (!app) {
    const credentialsPath = loadEnv('credentials');
    const databaseURL = loadEnv('urldbfirebase');

    let serviceAccount;
    // Resolve path from root
    const absolutePath = path.isAbsolute(credentialsPath)
      ? credentialsPath
      : path.join(process.cwd(), credentialsPath);

    if (fs.existsSync(absolutePath)) {
        serviceAccount = JSON.parse(fs.readFileSync(absolutePath, 'utf8'));
    } else {
        console.warn(`Credentials file not found at ${absolutePath}, attempting to initialize with default credentials.`);
    }

    const config = {
      databaseURL: databaseURL
    };
    if (serviceAccount) {
        config.credential = admin.credential.cert(serviceAccount);
    }

    app = admin.initializeApp(config);
  }
  return app;
}

async function getFirebaseData(tokenName) {
  console.log("=== GetFirebaseData Start ", tokenName);

  const cached = cacheRealtime.get(tokenName);
  if (cached) {
    console.log("=== GetFirebaseData Cache ", tokenName);
    return cached;
  }

  const db = getApp().database();
  const ref = db.ref("/" + tokenName);
  const snapshot = await ref.once('value');
  const data = snapshot.val();

  if (data) {
    let cacheTime = 1; // Default
    try {
      cacheTime = parseInt(loadEnv('CACHE_TIME_DATA'));
    } catch (e) {}

    cacheRealtime.set(tokenName, data, cacheTime * 60 * 60 * 1000); // Hours to ms
    return data;
  }

  throw new Error(`Not found token ${tokenName}`);
}

async function validateUser(email, uidUser) {
  const key = `${email}-${uidUser}`;
  const cached = cacheRealtime.get(key);
  if (cached !== null) {
    console.log("=== ValidateUser Cache ", key);
    return cached;
  }

  const db = getApp().database();
  const ref = db.ref("/users");

  // Use Firebase queries for efficiency
  const snapshot = await ref.orderByChild('email').equalTo(email).once('value');
  const data = snapshot.val();

  if (data) {
    const users = Object.values(data);
    for (const element of users) {
      if (element && element.email === email && element.uidUser === uidUser) {
        let cacheTime = 1;
        try {
          cacheTime = parseInt(loadEnv('CACHE_TIME_USER'));
        } catch (e) {}
        cacheRealtime.set(key, true, cacheTime * 60 * 60 * 1000);
        return true;
      }
    }
  }

  let cacheTimeFalse = 1;
  try {
    cacheTimeFalse = parseInt(loadEnv('CACHE_TIME_USER_FALSE'));
  } catch (e) {}
  cacheRealtime.set(key, false, cacheTimeFalse * 60 * 60 * 1000);
  return false;
}

module.exports = { getFirebaseData, validateUser, getApp };
